src/Domain/FunctionsOfTime/FunctionsOfTimeAreReady.hpp

```diff
@@ -36,55 +38,102 @@ class TaggedTuple;
-        std::array<std::string, 0>{}) {
-  const auto& proxy =
-      ::Parallel::get_parallel_component<Component>(cache)[array_index];
-  const Parallel::ArrayComponentId array_component_id =
-      Parallel::make_array_component_id<Component>(array_index);
+    const Component* component_p, const double time,
+    const std::optional<std::unordered_set<std::string>>& functions_to_check =
+        std::nullopt) {
+  using ProxyType = std::decay_t<decltype(
+      ::Parallel::get_parallel_component<Component>(cache)[array_index])>;
+  return detail::functions_of_time_are_ready_impl<
+      CacheTag, Parallel::PerformAlgorithmCallback<ProxyType>>(
+      cache, array_index, component_p, time, functions_to_check);
+}
 
-  return Parallel::mutable_cache_item_is_ready<CacheTag>(
-      cache, array_component_id,
-      [&functions_to_check, &proxy,
-       &time](const std::unordered_map<
-              std::string,
-              std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
-                  functions_of_time) {
-        using ::operator<<;
-        ASSERT(alg::all_of(
-                   functions_to_check,
-                   [&functions_of_time](const std::string& function_to_check) {
-                     return functions_of_time.count(function_to_check) == 1;
-                   }),
-               "Not all functions to check ("
-                   << functions_to_check << ") are in the global cache ("
-                   << keys_of(functions_of_time) << ")");
-        for (const auto& [name, f_of_t] : functions_of_time) {
-          if ((not functions_to_check.empty()) and
-              alg::find(functions_to_check, name) == functions_to_check.end()) {
-            continue;
-          }
-          const double expiration_time = f_of_t->time_bounds()[1];
-          if (time > expiration_time) {
-            return std::unique_ptr<Parallel::Callback>(
-                new Parallel::PerformAlgorithmCallback(proxy));
-          }
-        }
-        return std::unique_ptr<Parallel::Callback>{};
-      });
+/// \ingroup ComputationalDomainGroup
+/// Check that functions of time are up-to-date.
+///
+/// Check that functions of time in \p CacheTag with names in \p
+/// functions_to_check are ready at time \p time.  If  \p functions_to_check is
+/// a `std::nullopt`, checks all functions in \p CacheTag.  If any function is
+/// ready, schedules a `Parallel::SimpleActionCallback` with the GlobalCache
+/// which calls the simple action passed in as a template parameter. The `Args`
+/// are moved into the callback.
+template <typename CacheTag, typename SimpleAction, typename Metavariables,
+          typename ArrayIndex, typename Component, typename... Args>
+bool functions_of_time_are_ready(
+    Parallel::GlobalCache<Metavariables>& cache, const ArrayIndex& array_index,
+    const Component* component_p, const double time,
+    const std::optional<std::unordered_set<std::string>>& functions_to_check,
+    Args... args) {
```

Comment:
`Args&&` and forward below.

Reply:
I tried this first, but ran into build errors in `SimpleActionCallback` due to the `const &`ness of some of the args that were passed in. I wasn't sure how to get around that because `SimpleActionCallback` needs to actually store the data locally (not references to it which may go out of scope). That's why I chose a copy initially and then `std::move` from then on.

Reply:
That's a bug in `SimpleActionCallback`.  It shouldn't be taking it's arguments by reference, but by value and then moving.

Reply:
Ok added a new commit that fixes `SimpleActionCallback`